softmax/src/lib.rs

```diff
@@ -1,82 +1,191 @@
-pub struct Softmax {}
-
-impl Softmax {
-    pub const fn new() -> Self { Softmax {} }
+use crate::{
+    proc_block_v1::{BadInputReason, GraphError, InvalidInput, KernelError},
+    runtime_v1::{
+        register_node, supported_shapes, Dimensions, ElementType, GraphContext,
+        KernelContext, Metadata, TensorMetadata, TensorParam, TensorResult,
+    },
+};
+
+use hotg_rune_proc_blocks::{
+    ndarray, ndarray::ArrayViewMut1, BufferExt, ValueType,
+};
+use num_traits::Float;
+
+wit_bindgen_rust::import!("../wit-files/rune/runtime-v1.wit");
+wit_bindgen_rust::export!("../wit-files/rune/proc-block-v1.wit");
+
+struct ProcBlockV1;
+
+fn softmax<T>(mut input: ArrayViewMut1<'_, T>)
+where
+    T: Float + num_traits::FromPrimitive,
+{
+    input.mapv_inplace(|x| x.exp());
+
+    let sum = input.sum();
+    if !num_traits::Zero::is_zero(&sum) {
+        input.mapv_inplace(|x| x / sum);
+    }
 }
 
-impl Default for Softmax {
-    fn default() -> Self { Softmax::new() }
+fn preprocess_buffer<'buf, T>(
+    buffer: &'buf mut [u8],
+    dimensions: &[u32],
+) -> Result<ArrayViewMut1<'buf, T>, KernelError>
+where
+    T: ValueType,
+{
+    buffer
+        .view_mut::<T>(dimensions)
+        .and_then(|t| t.into_dimensionality())
+        .map_err(|e| {
+            KernelError::InvalidInput(InvalidInput {
+                name: "confidences".to_string(),
+                reason: BadInputReason::InvalidValue(e.to_string()),
+            })
+        })
 }
 
-impl Transform<Tensor<f32>> for Softmax {
-    type Output = Tensor<f32>;
+impl proc_block_v1::ProcBlockV1 for ProcBlockV1 {
+    fn register_metadata() {
+        let metadata = Metadata::new("Softmax", env!("CARGO_PKG_VERSION"));
+        metadata.set_description(env!("CARGO_PKG_DESCRIPTION"));
+        metadata.set_repository(env!("CARGO_PKG_REPOSITORY"));
+        metadata.set_homepage(env!("CARGO_PKG_HOMEPAGE"));
+        metadata.add_tag("softmax");
+        metadata.add_tag("image");
+        metadata.add_tag("nlp");
+        metadata.add_tag("numeric");
+        metadata.add_tag("classification");
+
+        let input = TensorMetadata::new("input");
+        let hint = supported_shapes(
+            &[ElementType::F32, ElementType::F64],
+            Dimensions::Fixed(&[0]),
+        );
+        input.add_hint(&hint);
+        metadata.add_input(&input);
+
+        let soft_max = TensorMetadata::new("soft_max");
+        soft_max
+            .set_description("Vector normalised into probability distribution");
+        let hint = supported_shapes(
+            &[ElementType::F32, ElementType::F64],
+            Dimensions::Fixed(&[0]),
+        );
+        soft_max.add_hint(&hint);
+        metadata.add_output(&soft_max);
+
+        register_node(&metadata);
+    }
+
+    fn graph(id: String) -> Result<(), GraphError> {
+        let ctx = GraphContext::for_node(&id).ok_or_else(|| {
+            GraphError::Other("Unable to get the graph context".to_string())
+        })?;
```

Comment:
We found we were handling this sort of error so frequently we added a specific `GraphError::MissingContext` variant for it.

```suggestion
        let ctx = GraphContext::for_node(&id).ok_or(GraphError::MissingContext)?;
```